README
README for RudeSession
See the end of file for copying conditions

This file contains the following sections, please
glance through them all before installing RudeSession.

	* Introduction
	* Bug Reporting
	* Copying
	

*** INTRODUCTION ***

RudeSession is a library that provides session services to web applications.

Documentation, discussions, and new releases are available at 

http://www.rudeserver.com/

See file `COPYING' for copying conditions.
See file `INSTALL' for compilation and installation instructions.
See file `NEWS' for a list of major changes in the current release.
See file `THANKS' for a list of contributors.
See file 'AUTHORS' for contact information


*** BUG REPORTING ***

Please Send bug reports to <rudesession@rudeserver.com>.
A bug report should contain an adequate description of the 
problem, your input, what you expected, what you got, 
and why this is wrong.  Where applicable, please submit actual data
files with your report. 

Your feedback will help us to make a better and more portable package.
Consider documentation errors as bugs, and report them as such.  If you
develop anything pertaining to RudeSession or have suggestions, let us know
and share your findings by writing to <rudesession@rudeserver.com> or
by visiting the discussion forum accessable from the RudeServer 
website at http://www.rudeserver.com/



*** COPYING ***

Copyright (C) 2005 Matthew Flood
See file AUTHORS for contact information

This file is part of RudeSession.

RudeSession is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

RudeSession is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with RudeSession; see the file COPYING.  If not, write to
the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.


NEWS
RudeSession NEWS - User visible changes.
See ChangeLog for more detailed information on internal changes
See the end for copying conditions.

__________________________________________________________________________________
Version 3.2.2-1
- Created MAN page
- Minor build script modifications

__________________________________________________________________________________
Version 3.2.1-1
- Added support for MySQL Storage of data

	Example:
	///////////////////////////////////////
	// Set up database context...
	//
	Database::addContext( "mycontextname", "mydatabase", "myusername", "mypassword", "serveraddress", 3306);

	// Set up session to use database
	//
	Session::setPersistanceMethod("database");
	Session::setDatabaseContext("mycontextname");

	// Use session as normal...
	//
	Session *session = session::getSession();
	//////////////////////////////////////////


	
	Database tables:
	
	CREATE TABLE IF NOT EXISTS `Sessions` (
  		`SessionID` char(50) NOT NULL,
  		`OriginalIP` char(15) NOT NULL default '',
  		`LastIP` char(15) NOT NULL default '',
  		`TimeCreated` datetime NOT NULL default '0000-00-00 00:00:00',
  		`TimeLastAccessed` datetime NOT NULL default '0000-00-00 00:00:00',
  		`MaxInactiveInterval` int(11) NOT NULL default '0',
  		`MaxSessionLength` int(11) NOT NULL default '0',
  		`NumTimesActive` int(10) unsigned NOT NULL default '0',
  		`IPMask` int(10) unsigned NOT NULL default '0',
  		PRIMARY KEY  (`SessionID`),
  		KEY `TimeCreated` (`TimeCreated`),
  		KEY `TimeLastAccessed` (`TimeLastAccessed`)
	) ENGINE=MyISAM DEFAULT CHARSET=latin1 |



	CREATE TABLE IF NOT EXISTS `SessionData` (
  		`SessionID` char(50) NOT NULL,
  		`DataName` varchar(255) NOT NULL default '',
  		`DataValue` mediumtext NOT NULL,
  		`ListOrder` int(11) NOT NULL default '0',
  		KEY `SessionID` (`SessionID`)
	) ENGINE=MyISAM DEFAULT CHARSET=latin1 

	
__________________________________________________________________________________
Version 3.1.1-1
- Changed entropy file from /dev/random to /dev/urandom
	because /dev/random was blocking on our new machine - no entropy was available!!!
	/dev/urandom does not block
__________________________________________________________________________________
Version 3.0.1-1

	** Detecting command line interactions and prompting user to create simple ID
	** Added debug features

	** Changed package versioning: 
		now using the LIBRARY_VERSION-RELEASE_VERSION
		Release version will be set to 1 for each new lib version
		
	** Rolled the package into binary and source RPMs
__________________________________________________________________________________
Version 2.1

	** Fixed a bug that allowed illegal characters to appear in session ID.

__________________________________________________________________________________
Version 2.0

	** RudeSession became open source and is distributed under the GNU Public License.
		Commercial licenses are available for a fee, and allow the holder
		to use the library in commercial applications. Contact info@rudeserver.com
		for more details.

__________________________________________________________________________________

Copyright (C) 2005 Matthew Flood
See file AUTHORS for contact information

This file is part of RudeSession.

RudeSession is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

RudeSession is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with RudeSession; see the file COPYING.  If not, write to
the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.



src/Implementation.h
// Implementation.h
//
// Copyright (C) 2005 Matthew Flood
// See file AUTHORS for contact information
//
// This file is part of RudeSession.
//
// RudeSession is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2, or (at your option)
// any later version.
// 
// RudeSession is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with RudeSession; (see COPYING) if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
// 02111-1307, USA.
//------------------------------------------------------------------------


#ifndef INCLUDED_Implementation_H
#define INCLUDED_Implementation_H


#include "MetaData.h"#include "DataStore.h"
#include <string>
#include <map>



namespace rude{
namespace session{
class AbstractPersistance;


class Implementation{

	static std::string s_error;
	bool d_isreadonly;
	MetaData d_metadata;
	DataStore d_datastore;
	AbstractPersistance *d_persistance;


protected:

	// done
	static void setError(const char *error);

public:

	Implementation();
	~Implementation();


	//////////////////
	// Global
	///////////////////

	// done
	static const char *getError();



	// TODO
	int purgeSessions();

	// done
	bool terminate();

	// done
	bool save();

	// done (should we save or just mark as modified?)
	bool update();

	// done (should we terminate expired?)
	bool validate();



bool createSession();
bool loadSession(const char *id);

	////////////////////
	// Must use instance
	/////////////////////


	bool generateNewSessionID();


	// done
	long getCreationTime();

	// done
	long getLastAccessedTime();

	// done
	const char *getOriginalIP();

	// done
	const char *getLastIP();

	// done
	int getNumTimesActive();

	// done
	const char *getId();

	// done
	void setMaxInactiveInterval(int minutes);

	// done
	int getMaxInactiveInterval();

	// done
	int getIPMask();

	// done
	void setIPMask(int maskbits);

	// done
	bool isNew();




////////////////////////////////
// Data Methods
////////////////////////////////

	bool exists(const char *name);

	int getNumValues();
	int getNumValues(const char *name);
	
	const char *getNameAt(int);
	const char *getValueAt(int);

	void setValue(const char *name, const char *value);
	const char *getValue(const char *name);

	void addValue(const char *name, const char *value);
	const char *getValue(const char *name, int index);

	void deleteValue(const char *name);
	const char * removeValue(const char *name, int index);

	void clear();
};

}}

#endif


src/Base64Encoder.cpp
// Base64Encoder.cpp
//
// Copyright (C) 2005 Matthew Flood
// See file AUTHORS for contact information
//
// This file is part of RudeSession.
//
// RudeSession is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2, or (at your option)
// any later version.
// 
// RudeSession is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with RudeSession; (see COPYING) if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
// 02111-1307, USA.
//------------------------------------------------------------------------



#include "Base64Encoder.h"

#ifndef INCLUDED_STRING_H
#include <string.h>
#define INCLUDED_STRING_H
#endif

namespace rude{
namespace session{

// translates a 6 bit char into a base-64 character
//
inline char Base64Encoder::c_encode(char uc)
{
	if (uc < 26)
	{
		return 'A'+uc;
	}
	if (uc < 52)
	{
		return 'a'+(uc-26);
	}
    if (uc < 62)
	{
		return '0'+(uc-52);
	}
	if (uc == 62)
	{
		return '+';
	}
	return '/';
}

// translates one base-64 character into a six-bit pattern
//
inline unsigned char Base64Encoder::c_decode(char c)
{
      if (c >= 'A' && c <= 'Z')
      {
            return c - 'A';
      }
      if (c >= 'a' && c <= 'z')
      {
            return c - 'a' + 26;
      }
      if (c >= '0' && c <= '9')
      {
            return c - '0' + 52;
      }
      if (c == '+')
      {
            return 62;
      };
      return 63;
}

char * Base64Encoder::encode(const char *data, int datalength)
{
      
	
      int linelength=0;
      int outlength=0;
 
      if (data == (char*) 0 || datalength == 0)
      {
            return (char*) 0;
      };

		char *retval=new char[datalength * 2];
		const char *crlf="\n";
		int crlflength=strlen(crlf);
		int maxlinelength=76;
		
      for (int i=0; i<datalength; i+=3)
      {
				unsigned char by1, by2, by3, by4, by5, by6, by7;
            by1 = data[i];
				
            if (i+1 < datalength)
            {
                  by2 = data[i+1];
            }
				else
				{
	            by2=0;
				}
				
            if (i+2 < datalength)
            {
                  by3 = data[i+2];
            }
				else
				{
					by3=0;
				}

            by4 = by1>>2;
            by5 = ((by1 & 0x3)<<4)|( by2>>4 );
            by6 = ((by2 & 0xf)<<2)|( by3>>6 );
            by7 = by3 & 0x3f;



            retval[outlength++] = c_encode(by4);
            linelength++;
				
            retval[outlength++] = c_encode(by5);
            linelength++;
 
            if (i+1 < datalength)
            {
                  retval[outlength++]= c_encode(by6);
                  linelength++;
            }
            else
            {
                  retval[outlength++]= '=';
                  linelength++;
            }

            if (i+2<datalength)
            {
                  retval[outlength++]= c_encode(by7);
                  linelength++;
            }
            else
            {
                  retval[outlength++]= '=';
                  linelength++;
            };

				// only set linelength if maxlinelength > 0
				//
            if (maxlinelength != 0 && linelength >= maxlinelength)
            {
						for(int y=0; y< crlflength; y++)
						{
							retval[outlength++]= crlf[y];
						}
						
						// we just created a line - reset current linelength
						//
                  linelength=0;
            }
      };
      retval[outlength]=(char) NULL;
      return retval;
}


char * Base64Encoder::decode(const char *data, int &outlength)
{
		// Error if incoming data is NULL
		// Error if nothing to decode
		//
		if(data == (char *) 0)
		{
			return (char*) 0;
		}

	   int datalength = strlen(data);
		if(!datalength)
		{
			return (char*) 0;
		}
		
      outlength=0;

		// rawlength is the length of the encoded data excluding
		// any non-base64 characters
		//
      int rawlength=0;


		// create return buffer
		//
		char *outbuffer=new char[datalength];

      // copy all base64 characters into outbuffer,
      // in other words, strip away CRLF's and non-b64 characters...
		//
      for (int j=0;j < datalength; j++)
      {
            if (IsBase64(data[j]))
            {
                  outbuffer[rawlength++]= data[j];
            }
      }

		// there's no base64 characters to decode
		//
      if (rawlength == 0 || datalength == 0)
      {
				delete [] outbuffer;
            return (char*) NULL;
      }

 

      for (int i=0; i<rawlength; i+=4)

      {

            

            char c1 = outbuffer[i];
				char c2='A',c3='A',c4='A';

            if (i+1 < rawlength)

            {

                  c2 = outbuffer[i+1];

            };

            if (i+2 < rawlength)

            {

                  c3 = outbuffer[i+2];

            };

            if (i+3 < rawlength)

            {

                  c4 = outbuffer[i+3];

            };

 

            unsigned char by1,by2,by3,by4;

            by1 = c_decode(c1);

            by2 = c_decode(c2);

            by3 = c_decode(c3);

            by4 = c_decode(c4);

 

            outbuffer[outlength]=( (by1<<2)|(by2>>4) );
            outlength++;
 

            if (c3 != '=')

            {

                  outbuffer[outlength]=( ((by2&0xf)<<4)|(by3>>2) );
                  outlength++;
            }

 

            if (c4 != '=')

            {

                  outbuffer[outlength]=( ((by3&0x3)<<6)|by4 );
                  outlength++;
            };

      };
		// NULL terminate decoded data
		// in case caller ignores outlength
		// for text data
		//
      outbuffer[outlength]=(char) 0;
      return outbuffer;
}

//The last helper function returns true is a character is a valid
//base-64 character and false otherwise.
inline bool Base64Encoder::IsBase64(char c)
{
      if (c >= 'A' && c <= 'Z')
      {
            return true;
      }
      if (c >= 'a' && c <= 'z')
      {
            return true;
      }
      if (c >= '0' && c <= '9')
      {
            return true;
      }
      if (c == '+')
      {
            return true;
      };
      if (c == '/')
      {
            return true;
      };
      if (c == '=')
      {
            return true;
      };
      return false;
}

}}



src/Implementation.cpp
// session.h
//
// Copyright (C) 2005 Matthew Flood
// See file AUTHORS for contact information
//
// This file is part of RudeSession.
//
// RudeSession is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2, or (at your option)
// any later version.
// 
// RudeSession is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with RudeSession; (see COPYING) if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
// 02111-1307, USA.
//------------------------------------------------------------------------


#include "Implementation.h"

#include "Configuration.h"
#include "IPValidator.h"
#include "IDGenerator.h"
#include "Cookie.h"
#include "CookieReader.h"

#include "AbstractPersistance.h"
#include "ConfigFilePersistance.h"
#include "DatabasePersistance.h"

//#define USING_VISUAL_STUDIO


#ifdef USING_VISUAL_STUDIO
#include <io.h>
#else
#ifndef INCLUDED_DIRENT_H
#include <dirent.h> // for unlink()
#define INCLUDED_DIRENT_H
#endif
#endif

#ifdef WIN32
#include <process.h> // for getpid
#else
#include <sys/types.h> // for getpid()
#include <unistd.h> // for getpid() and unlink()
#endif

#include <dirent.h>
#include <time.h>
#include <stdlib.h>
#include <stdio.h>
#include <iostream>
#include <string>
#include <string.h>
using namespace std;

namespace rude{
namespace session{
std::string Implementation::s_error="";


int Implementation::purgeSessions()
{
		if(! d_persistance->purgeSessions())
		{
			setError(d_persistance->getError());
			return false;
		}
		return true;
}


void Implementation::setError(const char *error)
{
	s_error = error ? error : "";
}

// CONSTRUCTOR & DESTRUCTOR
///////////////////////////////

Implementation::Implementation()
{
	d_isreadonly=false;

   bool debugmode = Configuration::getDebugMode();

	string persistanceType = Configuration::getPersistanceMethod();
	if(persistanceType == "database")
	{
		if(debugmode)
		{
			cerr << "SESSION DEBUG: " << __FILE__ << ":" << __LINE__ << " Creating Database Persistance Object\n";
		}
		d_persistance = new DatabasePersistance();
	}
	else
	{
		if(debugmode)
		{
			cerr << "SESSION DEBUG: " << __FILE__ << ":" << __LINE__ << " Creating Config File Persistance Object\n";
		}
		d_persistance = new ConfigFilePersistance();
	}


   string requestmethod=(getenv("REQUEST_METHOD") ? getenv("REQUEST_METHOD") : "");
   string request_id;
	
	if(requestmethod != "")
	{
	  // untaint supplied sid - it might contain .././../../..


	  if(debugmode)
	  {
	  	  cerr << "SESSION DEBUG: " << __FILE__ << ":" << __LINE__ << " Obtaining Requested Session ID from CookieReader using IDENTIFIER '" << Configuration::getCookieIdentifier() << "'\n";
	  }
 	  CookieReader cookie_reader;
	  request_id = cookie_reader.getValue(Configuration::getCookieIdentifier());
	}
	else
	{
	  if(debugmode)
	  {
	  	  cerr << "SESSION DEBUG: " << __FILE__ << ":" << __LINE__ << " Obtaining Requested Session ID from Command Line\n";
	  }
        cerr << "Session Command Line:  Please enter your session ID (x for none): ";
  		  std::getline(std::cin, request_id);
		  if(request_id == "x")
		  {
	        request_id="";
		  }
	}

	if(debugmode)
	{
	 	  cerr << "SESSION DEBUG: " << __FILE__ << ":" << __LINE__ << "Using ID: '" << request_id << "'\n";
	}
	d_metadata.setID(request_id.c_str());



	if(debugmode)
	{
		cerr << "SESSION DEBUG: " << __FILE__ << ":" << __LINE__ << " Attempting to Load and Validate Session\n";
	}
	if(request_id != "" && loadSession(request_id.c_str()) && validate())
	{
		if(debugmode)
		{
			cerr << "SESSION DEBUG: " << __FILE__ << ":" << __LINE__ << " Session Loaded and Validated - updating session meta data\n";
		}
		update();
	}
	else
	{
		if(debugmode)
		{
			cerr << "SESSION DEBUG: " << __FILE__ << ":" << __LINE__ << " Could not load session, creating new one\n";
		}
		createSession();
	}
}

bool Implementation::createSession()
{
   bool debugmode = Configuration::getDebugMode();
	
	// Comeup with an ID
	//
	int id_length=25;
	string requestmethod=(getenv("REQUEST_METHOD") ? getenv("REQUEST_METHOD") : "");
	string session_id;
	
	if(requestmethod == "")
	{
        cerr << "Session Command Line:  Creating Session -  please enter a session ID to use for future requests (g to generate): ";
  		  std::getline(std::cin, session_id);
		  if(session_id == "g")
		  {
		     cerr << "Generating ID\n";
	        session_id=IDGenerator::generateID(id_length);
		  }
	}
	else
	{
	      session_id=IDGenerator::generateID(id_length);
	}
	
	// Set the client cookie
	//
	Cookie c;
	c.setName(Configuration::getCookieIdentifier());
	c.setValue(session_id.c_str());
	c.setPath(Configuration::getPath());
	c.setDomain(Configuration::getDomain());
	c.setPort(Configuration::getPort());
	c.setDiscard(Configuration::getDiscard());
	c.setComment(Configuration::getComment());
	c.setCommentURL(Configuration::getCommentURL());
	c.setSecure(Configuration::getSecure());

	std::cout << c.cookieHeader();

	// The session might be be able to be created, for wahtever reason.
	// It might be argued to set ismodified to false so that the reason for failure
	// can be determined when explicitly calling save().
	// With this mindset, it would be tempting to only write out the cookie
	// after a successful save, but we have to do it in the creae method to ensure that
	// the set-cookie header acutally gets sent with the headers.
	// if we rely on a view to call save, or let the save be performed when the session
	// is destroyed, then the client may never get the cookie....
	//
	

		// Set the Meta Data
		//
		d_metadata.setID(session_id.c_str());
		
		const char *remoteaddr=getenv("REMOTE_ADDR");
		if(remoteaddr == (char*) 0 || strlen(remoteaddr) == 0)
		{
			remoteaddr= "127.0.0.1";
		}
		d_metadata.setOriginalIP(remoteaddr);
		d_metadata.setLastIP(remoteaddr);
		
		d_metadata.setMaxInactiveInterval(Configuration::getMaxInactiveInterval());
		d_metadata.setIPMask(Configuration::getDefaultIPMask());
		
		d_metadata.setCreationTime((long)time(0));
		d_metadata.setLastAccessedTime((long)time(0));
		d_metadata.setNumTimesActive(0);
		
		d_metadata.isNew(true);
	
		// Make sure there is no User Data
		//
		d_datastore.clear();

		if(d_persistance->createSession(d_metadata, d_datastore))
		{
			d_metadata.isModified(false);
			d_datastore.isModified(false);
			return true;
		}
		else
		{
				purgeSessions();
				d_metadata.isModified(true);
				d_datastore.isModified(true);
				setError(d_persistance->getError());
				return false;
		}
}

bool Implementation::generateNewSessionID()
{
	// This is a good place to purge sessions!!
	//

	purgeSessions();
		
	// Comeup with an ID
	//
	int id_length=25;
	string session_id=IDGenerator::generateID(id_length);
		
	// Set the client cookie
	//
	Cookie c;
	c.setName(Configuration::getCookieIdentifier());
	c.setValue(session_id.c_str());
	c.setPath(Configuration::getPath());
	c.setDomain(Configuration::getDomain());
	c.setPort(Configuration::getPort());
	c.setDiscard(Configuration::getDiscard());
	c.setComment(Configuration::getComment());
	c.setCommentURL(Configuration::getCommentURL());
	c.setSecure(Configuration::getSecure());

	std::cout << c.cookieHeader();

	// delete old session stuff
	//
	
	d_persistance->deleteSession(d_metadata.getId());

	// Set the new ID
	//
	d_metadata.setID(session_id.c_str());

	if(d_persistance->createSession(d_metadata, d_datastore))
	{
		d_metadata.isModified(false);
		d_datastore.isModified(false);
		return true;
	}
	else
	{
		d_metadata.isModified(true);
		d_datastore.isModified(true);
		setError(d_persistance->getError());
		return false;
	}
}	

bool Implementation::loadSession(const char *id)
{
		if( d_persistance->loadSession(d_metadata, d_datastore))
		{
			return true;
		}
		else
		{
			setError(d_persistance->getError());
			purgeSessions();
			return false;
		}
}

Implementation::~Implementation()
{
	save();
}

// INSTANCE METHODS

bool Implementation::terminate()
{
	// Clear cookies amd Session ID here????
	//
	Cookie c;
	c.setName(Configuration::getCookieIdentifier());
	c.setValue("");
	c.setPath(Configuration::getPath());
	c.setDomain(Configuration::getDomain());
	c.setPort(Configuration::getPort());
	c.setDiscard(Configuration::getDiscard());
	c.setComment(Configuration::getComment());
	c.setCommentURL(Configuration::getCommentURL());
	c.setSecure(Configuration::getSecure());
	
	cerr << c.cookieHeader();

	// Prevent resaving the file
	//
	d_metadata.isReadOnly(true);
	
	return true;
}

bool Implementation::save()
{
	if(!d_isreadonly)
	{
		if(d_metadata.isModified() || d_datastore.isModified())
		{
			if(! d_persistance->saveSession(d_metadata, d_datastore))
			{
				setError(d_persistance->getError());
				return false;
			}
			return true;
		}
		return true;
	}
	else
	{
		setError("Session is read only");
	}
	return false;
}

bool Implementation::update()
{
	if(!d_metadata.isNew())
	{
		d_metadata.update();
		return true;
	}
	else
	{
		setError("Cannot update new session");
		return false;
	}
}

bool Implementation::validate()
{
	const char *current=getenv("REMOTE_ADDR");
	
	if(!current)
	{
		current = "127.0.0.1";
	}

	if(IPValidator::validate(d_metadata.getOriginalIP(), current, d_metadata.getIPMask()))
	{
		long timenow = (long)time(0);

		// now validate the expiration thingy...
		//
		long interval = d_metadata.getMaxInactiveInterval();
		if(interval)
		{
			// convert minutes to seconds
			//
			interval *= 60;

			// timenow - lastactive cannot exceed interval
			//
			long difference = timenow - d_metadata.getLastAccessedTime();
			if(difference > interval)
			{
				setError("Session Timed Out");
				return false;
			}
		}

		// and validate session expiration
		// 
		long maxlength = d_metadata.getMaxSessionLength();
		if(maxlength)
		{
			// convert minutes to seconds
			//
			interval *= 60;

			// timenow - lastactive cannot exceed interval
			//
			long difference = timenow - d_metadata.getCreationTime();
			if(difference > maxlength)
			{
				setError("Session Expired");
				return false;
			}	
		}
		
		return true;
	}
	else
	{
		setError("Ip Mismatch");
		// Prevent saving any changes to session
		//
		return false;
	}
}

const char *Implementation::getError()
{
	return s_error.c_str();
}

//////////////////////////////////////////////////
/////  META-DATA METHODS
/////  ALL FUNCTIONS DEAL WITH OBJECT DATA-MEMBERS 
/////  AND UPDATE d_ismodified ACCORDINGLY
/////////////////////////////////////////////////

void Implementation::setMaxInactiveInterval(int minutes)
{
	d_metadata.setMaxInactiveInterval(minutes);
}

int Implementation::getMaxInactiveInterval()
{
	return d_metadata.getMaxInactiveInterval();
}

int Implementation::getIPMask()
{
	return d_metadata.getIPMask();
}

void Implementation::setIPMask(int maskbits)
{
	d_metadata.setIPMask(maskbits);
}

bool Implementation::isNew()
{
	return d_metadata.isNew();
}

long Implementation::getCreationTime()
{
	return d_metadata.getCreationTime();
}

long Implementation::getLastAccessedTime()
{
	return d_metadata.getLastAccessedTime();
}

const char *Implementation::getOriginalIP()
{
	return d_metadata.getOriginalIP();
}

const char *Implementation::getLastIP()
{
	return d_metadata.getLastIP();
}

int Implementation::getNumTimesActive()
{
	return d_metadata.getNumTimesActive();
}

const char *Implementation::getId()
{
	return d_metadata.getId();
}













//////////////////////////////////////////////////
/////  USER DATA METHODS
/////  ALL FUNCTIONS DEAL EXCLUSIVELY WITH DATA-MAP, 
/////  AND UPDATE d_ismodified ACCORDINGLY
/////////////////////////////////////////////////


void Implementation::setValue(const char *name, const char *value)
{
	d_datastore.setValue(name, value);
}


const char *Implementation::getValue(const char *name)
{
	return d_datastore.getValue(name);
}

void Implementation::addValue(const char *name, const char *value)
{
	d_datastore.addValue(name, value);
}

const char *Implementation::getValue(const char *name, int index)
{
	return d_datastore.getValue(name, index);
}

const char *  Implementation::removeValue(const char *name, int index)
{
	// Forward to implementation object
	//
	return d_datastore.removeValue(name, index);
}

int Implementation::getNumValues()
{
	return d_datastore.getNumValues();
}

int Implementation::getNumValues(const char *name)
{
	return d_datastore.getNumValues(name);
}

const char *Implementation::getNameAt(int index)
{
	return d_datastore.getNameAt(index);
}

const char *Implementation::getValueAt(int index)
{
	return d_datastore.getValueAt(index);
}

void Implementation::deleteValue(const char *name)
{
	d_datastore.deleteValue(name);
}

bool Implementation::exists(const char *name)
{
	return d_datastore.exists(name);
}

void Implementation::clear()
{
	d_datastore.clear();
}

}} // end namespaces

